README.md
# Electrode Hapi Compatibility Utility

A utility function that detects the Hapi version and return the appropriate plugin function.

Hapi 17 changed the signature of Plugins. This utility provides a simple wrapper for your plugin to support both Hapi 16 and Hapi 17+.

## Export plugin for Hapi 16 or 17+

If you have module that can export plugins for hapi 16 or 17+, you can use the API `universalHapiPlugin` to let this module automatically determine which of your plugins to use depending on the version of Hapi detected.

```js
const {universalHapiPlugin} = require("electrode-hapi-compat");

const registers = {
  hapi16: (server, options, next) => {...},
  hapi17OrUp: (server, options) => {...}
};

const pkg = {
  name: "MyPackage",
  version: "1.0.0"
};

module.exports = universalHapiPlugin(registers, pkg);
```

## Checking for Hapi 17 or Up

```js
const { isHapi17OrUp } = require("electrode-hapi-compat");

if (isHapi17OrUp()) {
  // hapi 17 or @hapi/hapi >= 18
} else {
  // hapi 16
}
```

## Checking for Hapi 18 or Up

```js
// this is to identify if @hapi/hapi v18 and above
const { isHapi18OrUp } = require("electrode-hapi-compat");

if (isHapi18OrUp()) {
  // @hapi/hapi >= 18
} else {
  // hapi 16/17
}
```

## Manually Setting Version for Testing

If you need to manually force a certain version of Hapi for testing etc,
you can manually set the Hapi major version this module should use with the `hapiVersion` property:

```js
// Set to use Hapi major version 18

require("electrode-hapi-compat").hapiVersion = 18;

// Get Hapi major version

const hapiVersion = require("electrode-hapi-compat").hapiVersion;
```

## Install

```bash
$ npm install --save electrode-hapi-compat
```

## Contribute

1. Clone this repo
2. Make updates
3. Run tests (requires 100% test coverage)
4. Submit PR
5. Sign CLA

To run tests

```bash
$ npm run test
```

## Publishing

1. Run `npm version <version-type>` to bump the version.
2. Update the `CHANGELOG.md`
3. Make sure your npmrc is pointing to external npm.
4. Run `npm publish` (make sure you have permission to github and npm).
5. Run `git push origin master --tags` to commit changes.

Built with :heart: by [Team Electrode](https://github.com/orgs/electrode-io/people) @WalmartLabs.

[hapi]: https://www.npmjs.com/package/hapi


CHANGELOG.md
## [v1.3.2] 01-10-2020

- Fix `_testSet` API edge case
- Instead of failing eagerly, fail on registration for fastify-incompatible plugins

## [v1.3.1] 01-02-2020

- Correct publish error

## [v1.3.0] 01-02-2020

- Support for fastify.

## [v1.2.0] 09-19-2019

- work with version directly instead of flags (#4)

## [v1.1.0] 09-13-2019

- update to handle @hapi/hapi@18

## [v1.0.0] 10-11-2018

- Initial release.


test/spec/util.spec.js
"use strict";

const hapi = require("hapi");
const mockRequire = require("mock-require");
const compat = require("../..");

describe("Util", () => {
  let index;
  let hapiServer;

  beforeEach(() => {
    hapiServer = new hapi.Server();
    hapiServer.connection({ port: 80 });
    hapiServer.route({
      method: "GET",
      path: "/test",
      handler: (req, reply) => {
        return reply("ok");
      }
    });
  });

  afterEach(() => {
    hapiServer.stop();
    delete require.cache[require.resolve("../..")];
    mockRequire.stopAll();
  });

  it("test index", () => {
    index = require("../..");
    expect(index.isHapi17()).false;
  });

  it("test is hapi 17", () => {
    mockRequire("hapi/package", { version: "17.2.2" });
    index = require("../..");
    expect(index.isHapi17()).true;
  });

  it("test is not hapi 17", () => {
    index = require("../..");
    expect(index.isHapi17()).false;
  });

  it("test allow tests to set isHapi17 flag", () => {
    index = require("../..");
    expect(index.isHapi17()).false;
    index._testSetHapi17(true);
    expect(index.isHapi17()).true;
  });

  it("test allow tests to set isFastify flag", () => {
    index = require("../..");
    expect(index.isFastify()).false;
    index._testSetFastify(true);
    expect(index.isFastify()).true;
  });

  it("test is hapi 18", () => {
    mockRequire("@hapi/hapi/package", { version: "18.3.2" });
    index = require("../..");
    expect(index.isHapi18OrUp()).true;
  });

  it("test is not hapi 18", () => {
    index = require("../..");
    expect(index.isHapi18OrUp()).false;
  });

  it("test is not fastify", () => {
    index = require("../..");
    expect(index.isFastify()).false;
  });

  it("test is fastify", () => {
    mockRequire("fastify/package", {});
    index = require("../..");
    expect(index.isFastify()).true;
  });

  it("test allow tests to set isHapi18 flag", () => {
    index = require("../..");
    expect(index.isHapi18OrUp()).false;
    index._testSetHapi18(true);
    expect(index.isHapi18OrUp()).true;
  });

  it("test no hapi, defaults hapi 16", () => {
    //mockRequire("@hapi/hapi/package", null);
    mockRequire("hapi/package", null);
    index = require("../..");
    expect(index.isHapi17()).false;
  });

  it("test universalHapiPlugin on Fastify", () => {
    mockRequire("fastify/package", {});
    index = require("../..");
    const registers = {
      hapi16: () => {},
      hapi17: () => {},
      fastify: () => {}
    };
    const pkg = { name: "Green" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    expect(plugin).equal(registers.fastify);
  });

  it("test universalHapiPlugin when no Fastify version", async () => {
    mockRequire("fastify/package", {});
    index = require("../..");
    const registers = {
      hapi16: () => {},
      hapi17: () => {}
    };
    const pkg = { name: "Green" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    try {
      await plugin();
      throw new Error("Expected plugin to throw on registration");
    } catch (e) {
      expect(e.message).to.equal("Plugin is not compatible with fastify");
    }
  });

  it("test universalHapiPlugin on Hapi 16", () => {
    index = require("../..");
    const registers = {
      hapi16: () => true,
      hapi17: () => false
    };
    const pkg = { name: "Green" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    expect(plugin.attributes.pkg).equal(pkg);
    expect(plugin).equal(registers.hapi16);
  });

  it("test universalHapiPlugin on Hapi 17", () => {
    mockRequire("hapi/package", { version: "17.0.0" });
    index = require("../..");
    const registers = {
      hapi16: () => false,
      hapi17: () => true
    };
    const pkg = { name: "Yellow" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    expect(plugin.pkg).equal(pkg);
    expect(plugin.register).equal(registers.hapi17);
  });

  it("test universalHapiPlugin on Hapi 18, using registers.hapi17", () => {
    mockRequire("hapi/package", { version: "18.3.2" });
    index = require("../..");
    const registers = {
      hapi16: () => false,
      hapi17: () => true
    };
    const pkg = { name: "Yellow" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    expect(plugin.pkg).equal(pkg);
    expect(plugin.register).equal(registers.hapi17);
  });

  it("test universalHapiPlugin on Hapi 18, using registers.hapi17OrUp", () => {
    mockRequire("hapi/package", { version: "18.3.2" });
    index = require("../..");
    const registers = {
      hapi16: () => false,
      hapi17OrUp: () => true
    };
    const pkg = { name: "Yellow" };
    const plugin = index.universalHapiPlugin(registers, pkg);
    expect(plugin.pkg).equal(pkg);
    expect(plugin.register).equal(registers.hapi17OrUp);
  });

  it("_testSetHapi17 should set version to 17 for true", () => {
    compat.hapiVersion = 18;
    compat._testSetHapi17(true);
    expect(compat.hapiVersion).equals(17);
  });

  it("_testSetHapi17 should set version to 16 for 17 and false", () => {
    compat.hapiVersion = 17;
    compat._testSetHapi17(false);
    expect(compat.hapiVersion).equals(16);
  });

  it("_testSetHapi17 should set version to 16 for 17 and false on fresh require", () => {
    mockRequire("hapi/package", { version: "17.8.5" });
    index = require("../..");
    index._testSetHapi17(false);
    expect(index.hapiVersion).equals(16);
  });

  it("_testSetHapi17 should do nothing for version !== 17 and false", () => {
    compat.hapiVersion = 18;
    compat._testSetHapi17(false);
    expect(compat.hapiVersion).equals(18);
  });

  it("_testSetHapi18 should set version to 18 for true", () => {
    compat.hapiVersion = 16;
    compat._testSetHapi18(true);
    expect(compat.hapiVersion).equals(18);
  });

  it("_testSetHapi18 should set version to 17 for 18 and false", () => {
    compat.hapiVersion = 18;
    compat._testSetHapi18(false);
    expect(compat.hapiVersion).equals(17);
  });

  it("_testSetHapi18 should set version to 17 for 18 and false on fresh require", () => {
    mockRequire("@hapi/hapi/package", { version: "18.3.2" });
    index = require("../..");
    index._testSetHapi18(false);
    expect(index.hapiVersion).equals(17);
  });

  it("_testSetHapi18 should do nothing for version !== 18 and false", () => {
    compat.hapiVersion = 16;
    compat._testSetHapi18(false);
    expect(compat.hapiVersion).equals(16);
  });
});


lib/util.js
"use strict";

/* eslint-disable no-magic-numbers */

const isValidObject = require("./is-valid-object");
const tryRequire = require("./try-require");

let HAPI_MAJOR_VERSION;
let FASTIFY;

function hapiMajor() {
  if (HAPI_MAJOR_VERSION === undefined) {
    HAPI_MAJOR_VERSION = 16;

    let hapiPkg = tryRequire("@hapi/hapi/package");
    if (!hapiPkg) {
      hapiPkg = tryRequire("hapi/package");
    }

    if (isValidObject(hapiPkg) && hapiPkg.version) {
      HAPI_MAJOR_VERSION = +hapiPkg.version.split(".")[0];
    }
  }

  return HAPI_MAJOR_VERSION;
}

function isHapi17() {
  return hapiMajor() >= 17;
}

function isHapi18OrUp() {
  return hapiMajor() >= 18;
}

function isFastify() {
  if (FASTIFY === undefined) {
    return (FASTIFY = Boolean(tryRequire("fastify/package")));
  }
  return FASTIFY;
}

function _testSetHapi17(isHapi17Flag) {
  if (isHapi17Flag) {
    HAPI_MAJOR_VERSION = 17;
  } else if (hapiMajor() === 17) {
    HAPI_MAJOR_VERSION = 16;
  }
}

function _testSetHapi18(isHapi18Flag) {
  if (isHapi18Flag) {
    HAPI_MAJOR_VERSION = 18;
  } else if (hapiMajor() === 18) {
    HAPI_MAJOR_VERSION = 17;
  }
}

function _testSetFastify(isFastifyFlag) {
  FASTIFY = isFastifyFlag;
}

/**
 * Detects the installed Hapi version and use the appropriate plugin
 *
 * @param {Object} registers: { hapi16, hapi17 } contains registers for hapi16, hapi17/18
 * @param {*|undefined} registers.fastify: optional field for fastify plugin
 * @param {*} pkg : package for a hapi plugin
 * @returns {*}: Hapi plugin appropriate for installed version
 *
 */
function universalHapiPlugin(registers, pkg) {
  if (isFastify()) {
    if (!registers.fastify) {
      // Lazily fail if the user tries to register this plugin with fastify.
      return () => Promise.reject(new Error("Plugin is not compatible with fastify"));
    }
    // allow modifying of parent fastify instance
    registers.fastify[Symbol.for("skip-override")] = true;
    return registers.fastify;
  } else if (hapiMajor() >= 17) {
    return {
      register: registers.hapi17OrUp || registers.hapi17,
      pkg
    };
  } else {
    registers.hapi16.attributes = {
      pkg
    };
    return registers.hapi16;
  }
}

module.exports = {
  hapiMajor,
  isHapi17OrUp: isHapi17,
  isHapi18OrUp,
  universalHapiPlugin,
  _testSetHapi17,
  _testSetHapi18,
  _testSetFastify,
  isHapi17,
  isFastify
};

Object.defineProperty(module.exports, "hapiVersion", {
  get: () => hapiMajor(),
  set: v => (HAPI_MAJOR_VERSION = v)
});
